File: Source/FlatShare/FlatshareCentral/src/main/java/com/joinflatshare/ui/profile/create/ProfileCreateListener.java
package com.joinflatshare.ui.profile.create;

import android.app.DatePickerDialog;
import android.view.View;

import androidx.core.content.ContextCompat;

import com.joinflatshare.FlatshareCentral.R;
import com.joinflatshare.FlatshareCentral.databinding.ActivityProfileCreateBinding;
import com.joinflatshare.customviews.alert.AlertDialog;
import com.joinflatshare.customviews.interests.InterestsView;
import com.joinflatshare.pojo.config.RentRange;
import com.joinflatshare.pojo.user.Name;
import com.joinflatshare.utils.helper.CommonMethods;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.regex.Pattern;

public class ProfileCreateListener implements View.OnClickListener {
    private final ProfileCreateActivity activity;
    private final ActivityProfileCreateBinding viewBind;
    private String gender = "", dob = "";

    public ProfileCreateListener(ProfileCreateActivity activity, ActivityProfileCreateBinding viewBind) {
        this.activity = activity;
        this.viewBind = viewBind;
        manageClicks();
    }

    private void manageClicks() {
        viewBind.cardProfileMale.setOnClickListener(this);
        viewBind.cardProfileFemale.setOnClickListener(this);
        viewBind.txtProfileDob.setOnClickListener(this);
        viewBind.btnProfileCreate.setOnClickListener(this);
        viewBind.llProfileInterests.setOnClickListener(this);
        viewBind.llProfileLanguages.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        int id = view.getId();
        if (id == R.id.card_profile_female) {
            viewBind.cardProfileFemale.setCardBackgroundColor(ContextCompat.getColor(activity, R.color.button_bg_black));
            viewBind.txtFemale.setTextColor(ContextCompat.getColor(activity, R.color.color_text_secondary));
            viewBind.cardProfileMale.setCardBackgroundColor(ContextCompat.getColor(activity, android.R.color.transparent));
            viewBind.txtMale.setTextColor(ContextCompat.getColor(activity, R.color.color_text_primary));
            viewBind.cardProfileMale.invalidate();
            viewBind.cardProfileFemale.invalidate();
            gender = "Female";
        } else if (id == R.id.card_profile_male) {
            viewBind.cardProfileMale.setCardBackgroundColor(ContextCompat.getColor(activity, R.color.button_bg_black));
            viewBind.txtMale.setTextColor(ContextCompat.getColor(activity, R.color.color_text_secondary));
            viewBind.cardProfileFemale.setCardBackgroundColor(ContextCompat.getColor(activity, android.R.color.transparent));
            viewBind.txtFemale.setTextColor(ContextCompat.getColor(activity, R.color.color_text_primary));
            viewBind.cardProfileMale.invalidate();
            viewBind.cardProfileFemale.invalidate();
            gender = "Male";
        } else if (id == R.id.txt_profile_dob) {
            final Calendar calendar = Calendar.getInstance();
            DatePickerDialog dpd = new DatePickerDialog(activity,
                    android.R.style.Theme_Holo_Dialog, (view12, year, month, dayOfMonth) -> {
                dob = String.format("%04d/%02d/%02d", year, (month + 1), dayOfMonth);
                String day = String.format("%02d/%02d/%04d", dayOfMonth, (month + 1), year);
                Calendar dobCalendar = Calendar.getInstance();
                dobCalendar.set(Calendar.YEAR, year);
                dobCalendar.set(Calendar.MONTH, month);
                dobCalendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
                calendar.add(Calendar.YEAR, -15);
                if (dobCalendar.getTime().after(calendar.getTime())) {
                    CommonMethods.makeToast("You need to be above the age of fifteen to join "
                            + activity.getString(R.string.app_name));
                    return;
                }
                viewBind.txtProfileDob.setText(day);
            }, calendar.get(Calendar.YEAR), calendar.get(Calendar.MONTH),
                    calendar.get(Calendar.DAY_OF_MONTH));
            dpd.getDatePicker().setMaxDate(calendar.getTimeInMillis());
            dpd.show();
        } else if (id == R.id.btn_profile_create) {
            if (viewBind.txtProfileFname.getText().toString().isEmpty())
                CommonMethods.makeToast("Please enter your first name");
            else if (viewBind.txtProfileFname.getText().toString().length() == 1)
                CommonMethods.makeToast("First name must be minimum 2 characters long");
            else if (viewBind.txtProfileLname.getText().toString().isEmpty())
                CommonMethods.makeToast("Please enter your last name");
            else if (viewBind.txtProfileLname.getText().toString().length() == 1)
                CommonMethods.makeToast("Last name must be minimum 2 characters long");
            else if (viewBind.txtProfileDob.getText().toString().isEmpty())
                CommonMethods.makeToast("Please enter your DOB");
            else if (gender.isEmpty())
                CommonMethods.makeToast("Please select your gender");
            else if (viewBind.txtProfileInterest.getText().toString().isEmpty())
                CommonMethods.makeToast("Please select your interests");
            else if (viewBind.txtProfileLanguages.getText().toString().isEmpty())
                CommonMethods.makeToast("Please select your languages");
            else {
                AlertDialog.showAlert(activity, "Heads Up", "You won't be able to edit your name, birthday, and gender. Is it confirmed?",
                        "Yes", "No", (intent, requestCode) -> {
                            if (requestCode == 1) {
                                Name name = new Name();
                                name.setFirstName(viewBind.txtProfileFname.getText().toString());
                                name.setLastName(viewBind.txtProfileLname.getText().toString());
                                activity.modelUser.setName(name);
                                activity.modelUser.setDob(dob);
                                activity.modelUser.setGender(gender);
                                activity.modelUser.getFlatProperties().setRentRange(new RentRange());
                                // interests
                                ArrayList<String> items = new ArrayList<>();
                                String text = viewBind.txtProfileInterest.getText().toString();
                                String[] split = text.split(Pattern.quote(", "));
                                if (split.length > 0) {
                                    for (String txt : split) {
                                        items.add(txt);
                                    }
                                }
                                activity.modelUser.getFlatProperties().setInterests(items);
                                // languages
                                items = new ArrayList<>();
                                text = viewBind.txtProfileLanguages.getText().toString();
                                split = text.split(Pattern.quote(", "));
                                if (split.length > 0) {
                                    for (String txt : split) {
                                        items.add(txt);
                                    }
                                }
                                activity.modelUser.getFlatProperties().setLanguages(items);
                                if (activity.getTemporaryLocation() == null)
                                    activity.checkLocation();
                                else {
                                    activity.modelUser.setLocation(activity.getTemporaryLocation());
                                    activity.setTemporaryLocation(null);
                                    activity.updateUser();
                                }
                            }
                        });
            }
        } else if (id == R.id.ll_profile_interests) {
            InterestsView interestsView = new InterestsView(activity, InterestsView.VIEW_TYPE_INTERESTS, viewBind.txtProfileInterest);
            interestsView.assignCallback(activity);
            interestsView.show();
        } else if (id == R.id.ll_profile_languages) {
            InterestsView interestsView;
            interestsView = new InterestsView(activity, InterestsView.VIEW_TYPE_LANGUAGES, viewBind.txtProfileLanguages);
            interestsView.assignCallback(activity);
            interestsView.show();
        }
    }
}

File: Flatshare Code/FlatShare 2.0/FlatshareCentral/src/main/java/com/joinflatshare/ui/register/create/ProfileCreateListener.kt
package com.joinflatshare.ui.register.create

import android.content.Intent
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.View.OnClickListener
import androidx.core.content.ContextCompat
import com.google.android.material.datepicker.CalendarConstraints
import com.google.android.material.datepicker.DateValidatorPointBackward
import com.google.android.material.datepicker.MaterialDatePicker
import com.joinflatshare.FlatshareCentral.R
import com.joinflatshare.FlatshareCentral.databinding.ActivityProfileCreateBinding
import com.joinflatshare.customviews.alert.AlertDialog
import com.joinflatshare.customviews.bottomsheet.BottomSheetView
import com.joinflatshare.customviews.bottomsheet.ModelBottomSheet
import com.joinflatshare.pojo.config.RentRange
import com.joinflatshare.pojo.user.Name
import com.joinflatshare.utils.helper.CommonMethod
import com.joinflatshare.utils.helper.CommonMethods
import com.joinflatshare.utils.helper.DateUtils
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale

/**
 * Created by debopam on 03/02/24
 */
class ProfileCreateListener(
    private val activity: ProfileCreateActivity,
    private val viewBind: ActivityProfileCreateBinding
) : OnClickListener {
    private var buttonListener: TextWatcher? = null
    var dob = ""

    init {
        viewBind.txtDob.setOnClickListener(this)
        viewBind.txtGender.setOnClickListener(this)
        viewBind.txtMyself.setOnClickListener(this)
        viewBind.btnBack.setOnClickListener(this)
        viewBind.btnCreateProfile.setOnClickListener(this)
        initTextWatcher()
        if (buttonListener != null) {
            viewBind.edtFname.addTextChangedListener(buttonListener)
            viewBind.edtLname.addTextChangedListener(buttonListener)
        }
    }

    override fun onClick(view: View?) {
        when (view?.id) {
            viewBind.txtGender.id -> {
                CommonMethods.hideSoftKeyboard(activity)
                val list = ArrayList<ModelBottomSheet>()
                list.add(ModelBottomSheet(0, "Male"))
                list.add(ModelBottomSheet(0, "Female"))
                list.add(ModelBottomSheet(0, "Trans"))
                BottomSheetView(activity, list) { _, position ->
                    viewBind.txtGender.text = list[position].name
                    checkValidation()
                }
            }

            viewBind.txtDob.id -> {
                val calendar = Calendar.getInstance()
                calendar.add(Calendar.YEAR, -15)
                val pickerNowTime = calendar.timeInMillis
                val constraints = CalendarConstraints.Builder().setEnd(pickerNowTime)
                    .setValidator(DateValidatorPointBackward.before(pickerNowTime))
                var sdf = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
                val dpd = MaterialDatePicker.Builder.datePicker().setTitleText("Select date")
                    .setSelection(calendar.timeInMillis)
                    .setTextInputFormat(sdf)
                    .setTheme(R.style.ThemeDatePicker)
                    .setCalendarConstraints(constraints.build())
                    .build()
                dpd.show(activity.supportFragmentManager, "DOB")
                dpd.addOnPositiveButtonClickListener { it ->
                    calendar.timeInMillis = it
                    viewBind.txtDob.text = sdf.format(calendar.time)
                    sdf = SimpleDateFormat("yyyy/MM/dd", Locale.getDefault())
                    dob = sdf.format(calendar.time)
                    checkValidation()
                }
            }

            viewBind.txtMyself.id -> {
                val list = ArrayList<ModelBottomSheet>()
                list.add(ModelBottomSheet(0, "Student"))
                list.add(ModelBottomSheet(0, "Working Professional"))
                BottomSheetView(activity, list) { _, position ->
                    viewBind.txtMyself.text = list[position].name
                    checkValidation()
                }
            }

            viewBind.btnBack.id -> {
                CommonMethod.finishActivity(activity)
            }

            viewBind.btnCreateProfile.id -> {
                if (checkValidation())
                    AlertDialog.showAlert(
                        activity,
                        "Alert",
                        "You won't be able to edit your name, birthday, & gender. Is it confirmed?",
                        "Yes, I know",
                        "No"
                    ) { _: Intent?, requestCode: Int ->
                        if (requestCode == 1) {
                            val name = Name()
                            name.firstName = viewBind.edtFname.text.toString()
                            name.lastName = viewBind.edtLname.text.toString()
                            val modelUser = activity.user
                            modelUser?.name = name
                            modelUser?.dob = dob
                            modelUser?.gender = viewBind.txtGender.text.toString()
                            modelUser?.flatProperties?.rentRange = RentRange()
                            modelUser?.profession = viewBind.txtMyself.text.toString()
                            activity.updateUser(modelUser)
                        }
                    }
            }
        }
    }

    private fun initTextWatcher() {
        buttonListener = object : TextWatcher {
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {

            }

            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {

            }

            override fun afterTextChanged(p0: Editable?) {
                checkValidation()

            }
        }
    }

    private fun checkValidation(): Boolean {
        var allFilled = true
        if (viewBind.edtFname.text.toString().trim().length < 2)
            allFilled = false
        else if (viewBind.edtLname.text.toString().trim().isEmpty())
            allFilled = false
        else if (viewBind.txtDob.text.toString().isEmpty())
            allFilled = false
        else if (viewBind.txtGender.text.toString().isEmpty())
            allFilled = false
        else if (viewBind.txtMyself.text.toString().isEmpty())
            allFilled = false
        if (allFilled)
            viewBind.btnCreateProfile.setBackgroundDrawable(
                ContextCompat.getDrawable(
                    activity,
                    R.drawable.drawable_button_blue
                )
            )
        else viewBind.btnCreateProfile.setBackgroundDrawable(
            ContextCompat.getDrawable(
                activity,
                R.drawable.drawable_button_light_blue
            )
        )
        return allFilled
    }

}